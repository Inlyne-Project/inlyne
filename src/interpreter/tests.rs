use std::{
    collections::VecDeque,
    path::PathBuf,
    sync::{
        atomic::{AtomicU32, Ordering},
        mpsc, Arc, Mutex,
    },
    thread,
    time::Duration,
};

use super::{HtmlInterpreter, ImageCallback, WindowInteractor};
use crate::{color::Theme, image::ImageData, Element, ImageCache};

use wgpu::TextureFormat;

// We use a dummy window with an internal counter that keeps track of when rendering a single md
// document is finished
#[derive(Default)]
struct AtomicCounter(Arc<AtomicU32>);

impl Clone for AtomicCounter {
    fn clone(&self) -> Self {
        Self(Arc::clone(&self.0))
    }
}

impl AtomicCounter {
    fn new() -> Self {
        Self(Arc::new(AtomicU32::from(1)))
    }

    fn is_finished(&self) -> bool {
        let counter = self.0.load(Ordering::SeqCst);
        counter == 0
    }

    fn inc(&self) {
        self.0.fetch_add(1, Ordering::SeqCst);
    }

    fn dec(&self) {
        self.0.fetch_sub(1, Ordering::SeqCst);
    }
}

struct DummyWindow(AtomicCounter);

impl WindowInteractor for DummyWindow {
    fn finished_single_doc(&self) {
        self.0.dec();
    }

    fn request_redraw(&self) {}

    // The counter is inc'd for each callback we create and internally dec's when it's called
    fn image_callback(&self) -> Box<dyn ImageCallback + Send> {
        self.0.inc();
        Box::new(DummyCallback(self.0.clone()))
    }
}

struct DummyCallback(AtomicCounter);

impl ImageCallback for DummyCallback {
    fn loaded_image(&self, _: String, _: Arc<Mutex<Option<ImageData>>>) {
        self.0.dec();
    }
}

fn dummy_interpreter(counter: AtomicCounter) -> (HtmlInterpreter, Arc<Mutex<VecDeque<Element>>>) {
    let element_queue = Arc::default();
    let theme = Theme::light_default();
    let surface_format = TextureFormat::Bgra8UnormSrgb;
    let hidpi_scale = 1.0;
    let file_path = PathBuf::from("does_not_exist");
    let image_cache = ImageCache::default();
    let window = Box::new(DummyWindow(counter));
    let interpreter = HtmlInterpreter::new_with_interactor(
        Arc::clone(&element_queue),
        theme,
        surface_format,
        hidpi_scale,
        file_path,
        image_cache,
        window,
    );

    (interpreter, element_queue)
}

fn interpret_md(text: &str) -> VecDeque<Element> {
    let counter = AtomicCounter::new();
    let (interpreter, element_queue) = dummy_interpreter(counter.clone());
    let (md_tx, md_rx) = mpsc::channel();
    md_tx.send(text.to_owned()).unwrap();
    let _ = std::thread::spawn(|| {
        interpreter.interpret_md(md_rx);
    });

    while !counter.is_finished() {
        thread::sleep(Duration::from_millis(1));
    }

    let mut elements_queue = element_queue.lock().unwrap();
    std::mem::take(&mut *elements_queue)
}

macro_rules! snapshot_interpreted_elements {
    ( $( ($test_name:ident, $md_text:ident) ),* $(,)? ) => {
        $(
            #[test]
            fn $test_name() {
                let text = $md_text;

                let syntect_theme = $crate::color::Theme::light_default().code_highlighter;
                let htmlified = $crate::utils::markdown_to_html(text, syntect_theme);
                let description = format!(" --- md\n\n{text}\n\n --- html\n\n{htmlified}");

                ::insta::with_settings!({
                    description => description,
                }, {
                    insta::assert_debug_snapshot!(interpret_md(text));
                });
            }
        )*
    }
}

const FOOTNOTES_LIST_PREFIX: &str = "\
This sentence[^1] has two footnotes[^2]

[^1]: 1st footnote
[^2]: 2nd footnote";

const CHECKLIST_HAS_NO_TEXT_PREFIX: &str = "\
- [x] Completed task
- [ ] Incomplete task";

snapshot_interpreted_elements!(
    (footnotes_list_prefix, FOOTNOTES_LIST_PREFIX),
    (checklist_has_no_text_prefix, CHECKLIST_HAS_NO_TEXT_PREFIX),
);
