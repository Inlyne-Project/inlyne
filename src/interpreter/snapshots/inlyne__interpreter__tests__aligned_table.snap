---
source: src/interpreter/tests.rs
description: " --- md\n\n| left default | left forced | centered | right | left default |\n| ------------ | :---------- | :------: | ----: | ------------ |\n| text         | text        |   text   |  text | text         |\n\n\n --- html\n\n<table>\n<thead>\n<tr>\n<th>left default</th>\n<th align=\"left\">left forced</th>\n<th align=\"center\">centered</th>\n<th align=\"right\">right</th>\n<th>left default</th>\n</tr>\n</thead>\n<tbody>\n<tr>\n<td>text</td>\n<td align=\"left\">text</td>\n<td align=\"center\">text</td>\n<td align=\"right\">text</td>\n<td>text</td>\n</tr>\n</tbody>\n</table>\n"
expression: "interpret_md_with_opts(text, opts)"
---
[
    Spacer(
        InvisibleSpacer(5),
    ),
    Table(
        Table {
            headers: [
                TextBox {
                    texts: [
                        Text {
                            text: "left default",
                            default_color: Color(BLACK),
                            style: BOLD ,
                            ..
                        },
                    ],
                    ..
                },
                TextBox {
                    texts: [
                        Text {
                            text: "left forced",
                            default_color: Color(BLACK),
                            style: BOLD ,
                            ..
                        },
                    ],
                    ..
                },
                TextBox {
                    align: Center,
                    texts: [
                        Text {
                            text: "centered",
                            default_color: Color(BLACK),
                            style: BOLD ,
                            ..
                        },
                    ],
                    ..
                },
                TextBox {
                    align: Right,
                    texts: [
                        Text {
                            text: "right",
                            default_color: Color(BLACK),
                            style: BOLD ,
                            ..
                        },
                    ],
                    ..
                },
                TextBox {
                    texts: [
                        Text {
                            text: "left default",
                            default_color: Color(BLACK),
                            style: BOLD ,
                            ..
                        },
                    ],
                    ..
                },
            ],
            rows: [
                [
                    TextBox {
                        texts: [
                            Text {
                                text: "text",
                                default_color: Color(BLACK),
                                ..
                            },
                        ],
                        ..
                    },
                    TextBox {
                        texts: [
                            Text {
                                text: "text",
                                default_color: Color(BLACK),
                                ..
                            },
                        ],
                        ..
                    },
                    TextBox {
                        align: Center,
                        texts: [
                            Text {
                                text: "text",
                                default_color: Color(BLACK),
                                ..
                            },
                        ],
                        ..
                    },
                    TextBox {
                        align: Right,
                        texts: [
                            Text {
                                text: "text",
                                default_color: Color(BLACK),
                                ..
                            },
                        ],
                        ..
                    },
                    TextBox {
                        texts: [
                            Text {
                                text: "text",
                                default_color: Color(BLACK),
                                ..
                            },
                        ],
                        ..
                    },
                ],
            ],
        },
    ),
    Spacer(
        InvisibleSpacer(5),
    ),
]
